### PR TITLE
Dam phase 2

### PR DESCRIPTION
**Pull request by eytannnaim**
Adding agent on boarder
and.. other refactors

### CODE REVIEW

File: examples/alpha/dam_basic_deployment/main.tf [hunk 1]
```diff
@@ -53,31 +52,57 @@ module "vpc" {
...
-  license_file        = var.license_file
-  key_pair            = module.key_pair.key_pair.key_pair_name
-  secure_password     = local.web_console_admin_password
-  mx_password         = local.web_console_admin_password
-  sg_ingress_cidr     = local.workstation_cidr
-  sg_ssh_cidr         = local.workstation_cidr
-  sg_web_console_cidr = local.workstation_cidr
-  attach_public_ip    = true
+  source                       = "../../../modules/aws/mx"
+  friendly_name                = join("-", [local.deployment_name_salted, "dam"])
```
Comment (lindanasredin) [inline examples/alpha/dam_basic_deployment/main.tf]: dam => mx

File: examples/alpha/dam_basic_deployment/main.tf [hunk 2]
```diff
@@ -53,31 +52,57 @@ module "vpc" {
...
-  key_pair               = module.key_pair.key_pair.key_pair_name
-  secure_password        = local.web_console_admin_password
-  mx_password            = local.web_console_admin_password
-  sg_ingress_cidr        = local.workstation_cidr
-  sg_agent_cidr          = var.agent_cidr_list
-  sg_ssh_cidr            = local.workstation_cidr
-  management_server_host = module.mx.private_ip
+  count                                   = var.gw_count
+  source                                  = "../../../modules/aws/agent-gw"
+  friendly_name                           = join("-", [local.deployment_name_salted, "dam", count.index])
```
Comment (lindanasredin) [inline examples/alpha/dam_basic_deployment/main.tf]: dam => agent-gw

File: examples/alpha/dam_basic_deployment/main.tf [hunk 3]
```diff
@@ -53,31 +52,57 @@ module "vpc" {
...
+resource "random_shuffle" "db" {
+  count = var.agents_count
+  input = ["PostgreSql", "MySql", "MariaDB"]
+}
+
+module "db_agent_monitored" {
+  count  = var.agents_count
+  source = "../../../modules/aws/db-agent-monitored"
+
+  friendly_name = join("-", [local.deployment_name_salted, "dam", count.index])
```
Comment (lindanasredin) [inline examples/alpha/dam_basic_deployment/main.tf]: dam => agent

File: examples/alpha/dam_basic_deployment/variables.tf [hunk 4]
```diff
@@ -20,6 +20,12 @@ variable "gw_count" {
   description = "Number of agent gateways"
 }
 
+variable "agents_count" {
```
Comment (lindanasredin) [inline examples/alpha/dam_basic_deployment/variables.tf]: agents_count => agent_count. Similar to gw_count

File: examples/alpha/dam_basic_deployment/variables.tf [hunk 5]
```diff
@@ -20,6 +20,12 @@ variable "gw_count" {
...
+  type        = number
+  default     = 1
+  description = "Number of databases with agent"
```
Comment (lindanasredin) [inline examples/alpha/dam_basic_deployment/variables.tf]: This description makes me think that there are databases without agent (which there is actually, the Sonar ones, but they are not present in this context, unless your context is all of DSF)

File: modules/aws/agent-gw/variables.tf [hunk 6]
```diff
@@ -1,6 +1,7 @@
 variable "friendly_name" {
   type        = string
   description = "Friendly name to identify all resources"
+  default     = "imperva-dsf-hub"
```
Comment (lindanasredin) [inline modules/aws/agent-gw/variables.tf]: imperva-dsf-hub => imperva-dsf-agent-gw

File: modules/aws/agent-gw/variables.tf [hunk 7]
```diff
@@ -137,15 +138,25 @@ variable "agent_listener_ssl" {
   default     = false
 }
 
-variable "management_server_host" {
+variable "management_server_host_for_registration" {
   type        = string
-  description = "Management server's hostname or IP address"
+  description = "Management server's hostname or IP address. Used for registering the Agent Gateway"
```
Comment (lindanasredin) [inline modules/aws/agent-gw/variables.tf]: Add period at the end.

File: modules/aws/agent-gw/variables.tf [hunk 8]
```diff
@@ -137,15 +138,25 @@ variable "agent_listener_ssl" {
...
   validation {
-    condition     = can(regex("[^0-9.]", var.management_server_host)) || cidrsubnet("${var.management_server_host}/32", 0, 0) == "${var.management_server_host}/32"
+    condition     = can(regex("[^0-9.]", var.management_server_host_for_registration)) || cidrsubnet("${var.management_server_host_for_registration}/32", 0, 0) == "${var.management_server_host_for_registration}/32"
     error_message = "Invalid IPv4 address"
   }
 }
 
+variable "management_server_host_for_api_access" {
+  type        = string
+  description = "Management server's hostname or IP address. It is utilized to access the API and ensure that the Agent Gateway is operational and ready. Leave empty if you wish to use the same value from management_server_host_for_registration variable"
```
Comment (lindanasredin) [inline modules/aws/agent-gw/variables.tf]: Add period at the end.

File: modules/aws/db-agent-monitored/variables.tf [hunk 9]
```diff
@@ -0,0 +1,74 @@
+variable "friendly_name" {
+  type        = string
```
Comment (lindanasredin) [inline modules/aws/db-agent-monitored/variables.tf]: default="imperva-dsf-agent-monitored-db"

File: modules/aws/db-agent-monitored/variables.tf [hunk 10]
```diff
@@ -0,0 +1,74 @@
...
+variable "sg_ssh_cidr" {
+  type        = list(string)
+  description = "List of allowed ingress CIDR patterns allowing ssh protocols to the ec2 instance"
+  default     = []
+}
+
+variable "db_type" {
+  type        = string
+  default     = "PostgreSql"
+  description = "DB type provision on ec2 with an agent, available types are - ['PostgreSql', 'MySql', 'MariaDB']"
```
Comment (lindanasredin) [inline modules/aws/db-agent-monitored/variables.tf]: Missing 'to" before provision. ec2 => EC2. Prefer : over -, for example, see error_message below.

File: modules/aws/db-agent-monitored/variables.tf [hunk 11]
```diff
@@ -0,0 +1,74 @@
...
+variable "registration_params" {
+  type = object(
+    {
+      agent_gateway_host = string
+      secure_password    = string
+      site               = string
+      server_group       = string
+    }
+  )
+  description = "Regisration parameters for DAM agent."
```
Comment (lindanasredin) [inline modules/aws/db-agent-monitored/variables.tf]: Remove period at the end.

File: modules/aws/mx/main.tf [hunk 12]
```diff
@@ -5,6 +5,7 @@ locals {
   required_udp_ports = []
   dam_model          = "AVM150"
   resource_type      = "mx"
+  mx_api_adderss     = module.mx.public_ip != null ? module.mx.public_ip : module.mx.private_ip
```
Comment (lindanasredin) [inline modules/aws/mx/main.tf]: The address is of the mx not the api. Maybe mx_ip_for_api

File: modules/aws/mx/variables.tf [hunk 13]
```diff
@@ -1,6 +1,7 @@
 variable "friendly_name" {
   type        = string
   description = "Friendly name to identify all resources"
+  default     = "imperva-dsf-hub"
```
Comment (lindanasredin) [inline modules/aws/mx/variables.tf]: imperva-dsf-mx

File: modules/aws/mx/variables.tf [hunk 14]
```diff
@@ -160,3 +161,9 @@ variable "large_scale_mode" {
   description = "Large scale mode"
   default     = false
 }
+
+variable "create_initial_configuration" {
+  type        = bool
+  description = "Create initial configuration to allow automatic agent on boarding"
```
Comment (lindanasredin) [inline modules/aws/mx/variables.tf]: on-boarding

File: modules/aws/agent-gw/README.md [hunk 15]
```diff
@@ -0,0 +1,91 @@
+# DSF DAM Agent Gateway
+[![GitHub tag](https://img.shields.io/github/v/tag/imperva/dsfkit.svg)](https://github.com/imperva/dsfkit/tags)
+
+This Terraform module provisions a DSF DAM Agent Gateway on AWS as an EC2 instance.
+
+The Database Monitoring Gateway performs application and database monitoring, providing full visibility into how that data is actually used in the enterprise, regardless of whether it is accessed directly or indirectly via applications.
```
Comment (lindanasredin) [inline modules/aws/agent-gw/README.md]: "The Database Monitoring Gateway" => "The Gateway". "that data" => "data"

File: modules/aws/agent-gw/README.md [hunk 16]
```diff
@@ -0,0 +1,91 @@
...
+
+## Requirements
+* Terraform v1.3.1.
```
Comment (lindanasredin) [inline modules/aws/agent-gw/README.md]: Terraform version between v1.3.1 and v1.4.x, inclusive

File: modules/aws/agent-gw/README.md [hunk 17]
```diff
@@ -0,0 +1,91 @@
...
+* An AWS account.
+* Network access to Management server (MX) on port 8083 (API and WebConsole)
+
+**NOTE:** In case you are not yet an Imperva customer, [please contact our team](https://www.imperva.com/contact-us/).
+
+## Resources Provisioned
+This Terraform module provisions several resources on AWS to create the DSF DAM Agent Gateway. These resources include:
+* An EC2 instance for running the software.
+* An EBS volume for storage.
+* A security groups to allow the required network access to and from the DSF instance.
```
Comment (lindanasredin) [inline modules/aws/agent-gw/README.md]: Remove the s in "security groups"

File: modules/aws/agent-gw/README.md [hunk 18]
```diff
@@ -0,0 +1,91 @@
...
+* `mx_password`: MX password
+* `secure_password`: The password used for communication between the Management Server and the Agent Gateway
+* `dam_version`: Version must be in the format dd.dd.dd.dd where each dd is a number between 1-99 (e.g 14.10.1.10)
+* `management_server_host_for_registration`: DSF DAM Management server address for registration
+
+Refer to [variables.tf](variables.tf) for additional variables with default values and additional info
+
+## Outputs
+
+The following [outputs](outputs.tf) are exported:
```
Comment (lindanasredin) [inline modules/aws/agent-gw/README.md]: exported meaning?

File: modules/aws/agent-gw/README.md [hunk 19]
```diff
@@ -0,0 +1,91 @@
...
+
+* `public_ip`: public address
```
Comment (lindanasredin) [inline modules/aws/agent-gw/README.md]: Capitalize the p in this and the following 3 lines.

File: modules/aws/agent-gw/README.md [hunk 20]
```diff
@@ -0,0 +1,91 @@
...
+* `private_ip`: private address
+* `public_dns`: public dns
+* `private_dns`: private dns
+* `display_name`: Display name of the instance under DSF portal
```
Comment (lindanasredin) [inline modules/aws/agent-gw/README.md]: I think it's better if we don't use the "DSF portal" term although it is valid and use the "DSF web console" term which we already use.

File: modules/aws/agent-gw/README.md [hunk 21]
```diff
@@ -0,0 +1,91 @@
...
+* `iam_role`: AWS IAM arn
+* `ssh_user`: SSH user for the instance
+* `instance_id`: AWS ec2 instance id
```
Comment (lindanasredin) [inline modules/aws/agent-gw/README.md]: ec2 => EC2

File: modules/aws/agent-gw/README.md [hunk 22]
```diff
@@ -0,0 +1,91 @@
...
+* `gateway_group_id`: The DSF Agent Group id of the instance
```
Comment (lindanasredin) [inline modules/aws/agent-gw/README.md]: id => Id. Is this the the name of the gateway group?

File: modules/aws/agent-gw/README.md [hunk 23]
```diff
@@ -0,0 +1,91 @@
...
+
+```
+module "dsf_agent_gw" {
+  source  = "imperva/dsf-agent-gw/aws"
+  version = "x.y.z"
+}
+```
+
+## API Access
+API access to the DSF DAM Management server is required to provision this module. Please make sure to pass the relevant CIDR block, representing your workstation, to allow such access through the `sg_ingress_cidr` variable
```
Comment (lindanasredin) [inline modules/aws/agent-gw/README.md]: Remove Please

File: modules/aws/agent-gw/README.md [hunk 24]
```diff
@@ -0,0 +1,91 @@
...
+
+## Additional Information
+
+For more information about the DSF Agent Gateway and its features, refer to the official documentation [here](https://docs.imperva.com/bundle/v14.11-database-activity-monitoring-user-guide/page/378.htm). 
+For additional information about DSF deployment using terraform, refer to the main repo README [here](https://github.com/imperva/dsfkit/tree/1.4.3).s
```
Comment (lindanasredin) [inline modules/aws/agent-gw/README.md]: Remove the 's' at the end of the line.

File: modules/aws/mx/README.md [hunk 25]
```diff
@@ -0,0 +1,91 @@
+# DSF DAM MX
+[![GitHub tag](https://img.shields.io/github/v/tag/imperva/dsfkit.svg)](https://github.com/imperva/dsfkit/tags)
+
+This Terraform module provisions a DSF DAM Management server (AKA mx) on AWS as an EC2 instance.
+
+Imperva’s Management Server Manager is a multi-domain management solution that dramatically improves the operational efficiency of managing SecureSphere deployments with multiple MX Management Servers. Management Server Manager meets the operational scalability demands of large enterprises and Managed Security Service Providers by consolidating the management, visibility and reporting across multiple SecureSphere MX Management Servers, and provides system wide health metrics and statistics.
```
Comment (lindanasredin) [inline modules/aws/mx/README.md]: This is a description of SOM, not MX.
I think it's better not to use a new term - "SecureSphere", we have enough terms, this is the first time we use it.

File: modules/aws/mx/README.md [hunk 26]
```diff
@@ -0,0 +1,91 @@
+# DSF DAM MX
+[![GitHub tag](https://img.shields.io/github/v/tag/imperva/dsfkit.svg)](https://github.com/imperva/dsfkit/tags)
+
+This Terraform module provisions a DSF DAM Management server (AKA mx) on AWS as an EC2 instance.
+
+Imperva’s Management Server Manager is a multi-domain management solution that dramatically improves the operational efficiency of managing SecureSphere deployments with multiple MX Management Servers. Management Server Manager meets the operational scalability demands of large enterprises and Managed Security Service Providers by consolidating the management, visibility and reporting across multiple SecureSphere MX Management Servers, and provides system wide health metrics and statistics. 
+
+## Requirements
+* Terraform v1.3.1.
```
Comment (lindanasredin) [inline modules/aws/mx/README.md]: Terraform version between v1.3.1 and v1.4.x, inclusive

File: modules/aws/mx/README.md [hunk 27]
```diff
@@ -0,0 +1,91 @@
...
+* An AWS account.
+* Network access to port 8083 (API and WebConsole)
+
+**NOTE:** In case you are not yet an Imperva customer, [please contact our team](https://www.imperva.com/contact-us/).
+
+## Resources Provisioned
+This Terraform module provisions several resources on AWS to create the DSF DAM MX. These resources include:
+* An EC2 instance for running the software.
+* An EBS volume for storage.
+* A security groups to allow the required network access to and from the DSF instance.
```
Comment (lindanasredin) [inline modules/aws/mx/README.md]: Remove the s in "security groups"

File: modules/aws/mx/README.md [hunk 28]
```diff
@@ -0,0 +1,91 @@
...
+The following input variables are **required**:
+
+* `subnet_id`: The ID of the subnet in which to launch the DSF instance
+* `key_pair`: AWS key pair name to attach to the instance
+* `mx_password`: MX password
+* `secure_password`: The password used for communication between the Management Server and the Agent Gateway
+* `dam_version`: Version must be in the format dd.dd.dd.dd where each dd is a number between 1-99 (e.g 14.10.1.10)
+* `license_file`: DAM license file path. Make sure this license is valid before deploying DAM otherwise this will result in an invalid deployment and loss of time
+
+Refer to [variables.tf](variables.tf) for additional variables with default values and additional info
```
Comment (lindanasredin) [inline modules/aws/mx/README.md]: Add period at the end.

File: modules/aws/agent-gw/README.md [hunk 29]
```diff
@@ -0,0 +1,91 @@
...
+The following input variables are **required**:
+
+* `subnet_id`: The ID of the subnet in which to launch the DSF instance
+* `key_pair`: AWS key pair name to attach to the instance
+* `mx_password`: MX password
+* `secure_password`: The password used for communication between the Management Server and the Agent Gateway
+* `dam_version`: Version must be in the format dd.dd.dd.dd where each dd is a number between 1-99 (e.g 14.10.1.10)
+* `management_server_host_for_registration`: DSF DAM Management server address for registration
+
+Refer to [variables.tf](variables.tf) for additional variables with default values and additional info
```
Comment (lindanasredin) [inline modules/aws/agent-gw/README.md]: Add period at the end.

File: modules/aws/mx/README.md [hunk 30]
```diff
@@ -0,0 +1,91 @@
...
+
+## Outputs
+
+The following [outputs](outputs.tf) are exported:
+
+* `public_ip`: public address
```
Comment (lindanasredin) [inline modules/aws/mx/README.md:43]: Capitalize the p in this and the following 3 lines.

File: modules/aws/mx/README.md [hunk 31]
```diff
@@ -0,0 +1,91 @@
...
+* `private_ip`: private address
+* `public_dns`: public dns
+* `private_dns`: private dns
+* `display_name`: Display name of the instance under DSF portal
+* `iam_role`: AWS IAM arn
+* `ssh_user`: SSH user for the instance
+* `instance_id`: AWS ec2 instance id
+* `configuration`: Pre configured site and service group available for use
```
Comment (lindanasredin) [inline modules/aws/mx/README.md]: Pre-configured

File: modules/aws/mx/README.md [hunk 32]
```diff
@@ -0,0 +1,91 @@
...
+* `mx_password`: MX password
+* `secure_password`: The password used for communication between the Management Server and the Agent Gateway
+* `dam_version`: Version must be in the format dd.dd.dd.dd where each dd is a number between 1-99 (e.g 14.10.1.10)
+* `license_file`: DAM license file path. Make sure this license is valid before deploying DAM otherwise this will result in an invalid deployment and loss of time
+
+Refer to [variables.tf](variables.tf) for additional variables with default values and additional info
+
+## Outputs
+
+The following [outputs](outputs.tf) are exported:
```
Comment (lindanasredin) [inline modules/aws/mx/README.md]: Again, what does exported here mean?

File: modules/aws/mx/README.md [hunk 33]
```diff
@@ -0,0 +1,91 @@
...
+}
+```
+
+## API Access
+API access is required to provision this module. Please make sure to pass the relevant CIDR block, representing your workstation, to allow such access through the `sg_ingress_cidr` variable
+
+## Additional Information
+
+For more information about the DSF MX and its features, refer to the official documentation [here](https://docs.imperva.com/bundle/v14.11-dam-management-server-manager-user-guide/page/10068.htm). 
+For additional information about DSF deployment using terraform, refer to the main repo README [here](https://github.com/imperva/dsfkit/tree/1.4.3).s
```
Comment (lindanasredin) [inline modules/aws/mx/README.md]: Remove the "s" at the end of the line.

**Review by lindanasredin — APPROVED** [review]

File: modules/aws/agent-gw/README.md (hunk 23 above)
Comment (eytannnaim) [reply modules/aws/agent-gw/README.md]: Why?

**Pull request merged**